Yoast/wpseo-news
# P1-343 Add JS translations

## Context


* Add JS translations. Meaning:
  * Retrieve wanted translation strings from our JS code.
  * Bundle those wanted translations with the ones from our PHP code, in one `.pot` file. That will be uploaded to our translators.
  * Download the actual translations, now also in JSON format for use in JS.
  * Process those JSON files to only contain the translations from JS.
  * Clean up all but plugin pot file (all wanted translations), mo files (actual translations for PHP) and json files (actual translations for JS).
  * Offer way to localize the language file for JS. Plus do so for existing JS script.

## Summary


This PR can be summarized in the following changelog entry:

* Adds translations for JavaScript.

## Relevant technical choices:

* Remove language files from Git, except for the main `.pot` file. This is kept to be able to easily see whether there are translation changes or not, when creating RCs.
* Ran into a slight problem because there are no JS translations in this plugin yet. Decided to add a dummy pot file to fake it being generated. This seems easier than adding safety checks in the rest of the pipeline.
* Deprecate `WPSEO_News_Javascript_Strings::strings` in favor of non-static.
* Preemptively add `@wordpress/i18n` as external and dependency.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Test translation building
* Run `grunt build:i18n`.
* Verify this updated the `languages/wordpress-seo-news.pot` file.
* Verify this created `.mo` and `.json` files for multiple languages in the `languages` folder.

#### Test translations being copied to artifact
* Run `grunt artifact`
* Verify the `artifact/languages` folder contains all the `.mo` and `.json` files (and the `index.php`).

#### Test JS translation detection
* There are currently no translations in JS. So let's create some to test.
* Add translations to the `js/src/yoast-seo-news-plugin.js` file.
  * The import on top: `import { __ } from "@wordpress/i18n";`
  * Inside the function: `document.getElementById( "wpseo_content_top" ).append( __( "Good news everyone!", "wordpress-seo-news" ) );` 
* Run `grunt build:i18n`
* Verify that the `languages/wordpress-seo-newsjs.json` file contains the string(s) you are 'translating' in the JS file.
* There are no actual translations yet, so the JSON files with locales don't include this (yet).

#### Test JS translations made available
* Build the plugin: `composer install && yarn && grunt build`
* Go to the News SEO settings: SEO -> News SEO
* Verify there are no JS errors on the page.
* Verify that `wpseoNewsJSL10n` is available on the window (check in the JS console).
  * When on English it will only have `{ wordpress-news-seo: null }`, but that is expected.
  * When on any other language it will have some empty object as value.
* Verify you see `Good news everyone!` at the bottom of the page.
* Switch your site language to something other than the default English.
* Add the actual translation to the JSON file of your chosen language.
  * It should be a sibling to the `"":{}` part and still be valid JSON.
  * NL example: `,"Good news everyone!":["Goed nieuws iedereen!"]`.
* Go to the News SEO settings: SEO -> News SEO
* Verify you see your translation at the bottom of the page, instead of the English default.

### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

* Not really testable yet until we have JS translations.
* Maybe go to SEO -> News SEO to verify no JS errors.

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes


## Documentation


## Quality assurance


Fixes https://yoast.atlassian.net/browse/P1-343
